Makefile

```diff
@@ -26,70 +26,113 @@ MAKEFLAGS += -j$(shell nproc)
+MLOS_VIZ_CONF_FILES := mlos_viz/setup.py mlos_viz/MANIFEST.in
+MLOS_GLOBAL_CONF_FILES := setup.cfg
+
+MLOS_PKGS := mlos_core mlos_bench mlos_viz
+MLOS_PKG_CONF_FILES := $(MLOS_CORE_CONF_FILES) $(MLOS_BENCH_CONF_FILES) $(MLOS_VIZ_CONF_FILES) $(MLOS_GLOBAL_CONF_FILES)
+
+build/conda-env.${CONDA_ENV_NAME}.build-stamp: ${ENV_YML} $(MLOS_PKG_CONF_FILES)
 	@echo "CONDA_SOLVER: ${CONDA_SOLVER}"
 	@echo "CONDA_EXPERIMENTAL_SOLVER: ${CONDA_EXPERIMENTAL_SOLVER}"
 	@echo "CONDA_INFO_LEVEL: ${CONDA_INFO_LEVEL}"
 	conda env list -q | grep -q "^${CONDA_ENV_NAME} " || conda env create ${CONDA_INFO_LEVEL} -n ${CONDA_ENV_NAME} -f ${ENV_YML}
 	conda env update ${CONDA_INFO_LEVEL} -n ${CONDA_ENV_NAME} --prune -f ${ENV_YML}
-	$(MAKE) clean-check clean-test clean-doc
+	$(MAKE) clean-format clean-check clean-test clean-doc clean-dist
 	touch $@
 
 .PHONY: clean-conda-env
 clean-conda-env:
 	conda env remove -y ${CONDA_INFO_LEVEL} -n ${CONDA_ENV_NAME}
 	rm -f build/conda-env.${CONDA_ENV_NAME}.build-stamp
 
+
+# Since these targets potentially change the files we need to run them in sequence.
+# In future versions of make we can do that by marking each as a .NOTPARALLEL psuedo target.
+# But with make 4.3 that changes the entire Makefile to be serial.
+
+# Here we make dynamic prereqs to apply to other targets that need to run in sequence.
+FORMAT_PREREQS :=
+
+.PHONY: format
+format: build/format.${CONDA_ENV_NAME}.build-stamp
+
+ifneq (,$(filter format,$(MAKECMDGOALS)))
+    FORMAT_PREREQS += build/format.${CONDA_ENV_NAME}.build-stamp
+endif
+
+build/format.${CONDA_ENV_NAME}.build-stamp: build/licenseheaders.${CONDA_ENV_NAME}.build-stamp
+# TODO: add isort and black formatters
+build/format.${CONDA_ENV_NAME}.build-stamp:
+	touch $@
+
+.PHONY: licenseheaders
+licenseheaders: build/licenseheaders.${CONDA_ENV_NAME}.build-stamp
+
+ifneq (,$(filter licenseheaders,$(MAKECMDGOALS)))
+    FORMAT_PREREQS += build/licenseheaders.${CONDA_ENV_NAME}.build-stamp
+endif
+
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp: build/conda-env.${CONDA_ENV_NAME}.build-stamp
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp: $(PYTHON_FILES)
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp: $(SCRIPT_FILES)
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp: $(SQL_FILES) doc/mit-license.tmpl
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp: doc/mit-license.tmpl
+build/licenseheaders.${CONDA_ENV_NAME}.build-stamp:
+	# Note: to avoid makefile dependency loops, we don't touch the setup.py
+	# files as that would force the conda-env to be rebuilt.
+	conda run -n ${CONDA_ENV_NAME} licenseheaders -t doc/mit-license.tmpl \
+		-E .py .sh .ps1 .sql .cmd \
+		-x mlos_bench/setup.py mlos_core/setup.py mlos_viz/setup.py
+	touch $@
+
+
 .PHONY: check
-check: pycodestyle pydocstyle pylint mypy # cspell licenseheaders markdown-link-check
+check: pycodestyle pydocstyle pylint mypy # cspell markdown-link-check
+# TODO: add isort and black checks
 
 .PHONY: pycodestyle
-pycodestyle: conda-env
```

Comment:
Unnecessary call to a PHONY target - we depend on the build-stamp files instead since that can be used to trigger rebuilds more precisely and avoid unnecessary rebuilds of a target.